### PR TITLE
support min_p sampling parameter

### PR DESCRIPTION
**Pull request by irexyc**
support min_p sampling option

https://github.com/InternLM/lmdeploy/issues/1745

### CODE REVIEW

File: src/turbomind/models/llama/LlamaV2.cc [hunk 1]
```diff
@@ -360,13 +360,8 @@ void LlamaV2<T>::dynamicDecode(int*            token_ids,
         {"local_batch_size", {MEMORY_CPU, TYPE_INT32, {1}, &local_batch_size}},
     };
 
-    const std::vector<std::string> optional_inputs{"stop_words_list",
-                                                   "bad_words_list",
-                                                   "runtime_top_k",
-                                                   "runtime_top_p",
-                                                   "temperature",
-                                                   "repetition_penalty",
-                                                   "random_seed"};
```
Comment (lvhan028) [inline src/turbomind/models/llama/LlamaV2.cc:369]: why remove `random_seed`

File: src/turbomind/models/llama/LlamaBatch.h [hunk 2]
```diff
@@ -3,7 +3,6 @@
 #pragma once
 
 // #include "src/turbomind/models/llama/LlamaCacheManager.h"
```
Comment (lvhan028) [inline src/turbomind/models/llama/LlamaBatch.h:5]: Let's remove this line

File: src/turbomind/layers/sampling_layers/StopCriteriaLayer.h [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "src/turbomind/layers/DynamicDecodeBaseLayer.h"
+#include "src/turbomind/macro.h"
+#include <vector>
```
Comment (lvhan028) [inline src/turbomind/layers/sampling_layers/StopCriteriaLayer.h:21]: no std::vector is used in `class StopCriteriaLayer`

File: src/turbomind/layers/DynamicDecodeBaseLayer.h [hunk 4]
```diff
@@ -23,27 +23,29 @@
 
 namespace turbomind {
 
+struct DynamicDecodeCommonArgs {
```
Comment (lvhan028) [inline src/turbomind/layers/DynamicDecodeBaseLayer.h:26]: Like `struct AttentionParam`, `struct ModelParam`, let's use `struct DecodeParam`

File: tests/csrc/unittests/test_sampling_layer.cu [hunk 5]
```diff
@@ -855,6 +854,7 @@ TYPED_TEST(SamplingDecodeTest, InvalidArgsBatchTopKBatchTopPContainZero)
     delete[] top_ps;
 }
 
+#if 0
```
Comment (lvhan028) [inline tests/csrc/unittests/test_sampling_layer.cu:857]: Why remove it?

File: tests/csrc/unittests/test_sampling_layer.cu [hunk 6]
```diff
@@ -1260,6 +1261,7 @@ TYPED_TEST(SamplingDecodeTest2, CorrectnessBatchRandTopP)
     this->runCurandTest({113, 1201, 1, 0, 1.0f, 5}, false, false);
 }
 
+#if 0
```
Comment (lvhan028) [inline tests/csrc/unittests/test_sampling_layer.cu:1264]: We still need to test topk sampling

File: src/turbomind/kernels/sampling_topp_kernels.h [hunk 7]
```diff
@@ -19,135 +19,62 @@
...
-                        int*            begin_offset_buf,
-                        curandState_t*  curandstate,
-                        const int       batch_size,
-                        const size_t    vocab_size_padded,
-                        const int*      end_ids,
-                        const float     top_p,
-                        cudaStream_t    stream,
-                        cudaDeviceProp* cuda_device_prop,
-                        const bool*     skip_decode);
+void invokeSoftMax(T* logits, const int vocab_size_padded, const int batch_size, const int* kept, cudaStream_t stream);
```
Comment (lvhan028) [inline src/turbomind/kernels/sampling_topp_kernels.h:32]: "SoftMax" -> "Softmax" 😂

File: src/turbomind/kernels/sampling_topp_kernels.h [hunk 8]
```diff
@@ -19,135 +19,62 @@
...
 };
 
-struct TopKPerSegmentContext {
-    TopKPerSegmentContext(): sm_count(0), sm_shared_size(0), sm_version(0){};
-    int sm_count;
-    int sm_shared_size;
-    int sm_version;
+struct TopPSortParams {
+    void*  workspace{};
+    size_t workspace_size{};
```
Comment (lvhan028) [inline src/turbomind/kernels/sampling_topp_kernels.h:51]: some members use in-place initialization while some don't. 
Any special reason for doing that?

File: src/turbomind/kernels/sampling_topp_kernels.h [hunk 9]
```diff
@@ -19,135 +19,62 @@
...
-    int num_items;
-    int num_segments;
-    // top_k per segment
-    int   num_top_k;
-    float top_p;
-    float confidence_threshold;
+template<typename T>
+void invokeTopPSort(TopPSortParams& params, cudaStream_t stream);
+
+struct TopPMinPFilterParams {
```
Comment (lvhan028) [inline src/turbomind/kernels/sampling_topp_kernels.h:66]: `struct TopPMinPFilterParams` is a subset of `struct TopPSortPrams` except the `float* min_ps`. 
Can we merge them together?

File: src/turbomind/layers/DynamicDecodeLayer.cc [hunk 10]
```diff
@@ -17,8 +17,9 @@
 #include "src/turbomind/layers/DynamicDecodeLayer.h"
 #include "src/turbomind/kernels/ban_bad_words.h"
```
Comment (lvhan028) [inline src/turbomind/layers/DynamicDecodeLayer.cc:18]: "ban_bad_words.h" and "stop_criteria_kernels.h" could be removed

File: tests/csrc/unittests/test_sampling_layer.cu (hunk 5 above)
Comment (irexyc) [reply tests/csrc/unittests/test_sampling_layer.cu]: The dynamic decoder doesn't support process mini-batch anymore

File: tests/csrc/unittests/test_sampling_layer.cu (hunk 6 above)
Comment (irexyc) [reply tests/csrc/unittests/test_sampling_layer.cu]: The dynamic decoder doesn't support process mini-batch anymore

File: src/turbomind/models/llama/LlamaV2.cc (hunk 1 above)
Comment (irexyc) [reply src/turbomind/models/llama/LlamaV2.cc:369]: “random_seed” has no use in dynamic decoder

File: lmdeploy/pytorch/engine/logits_process.py [hunk 11]
```diff
@@ -88,6 +88,22 @@ def _filter_topp_sorted(scores: torch.Tensor,
     return scores
 
 
+def _filter_minp_sorted(scores: torch.Tensor,
+                        minp: torch.Tensor,
+                        filter_value: float = -float('inf'),
+                        inplace: bool = True):
```
Comment (lvhan028) [inline lmdeploy/pytorch/engine/logits_process.py:94]: What's the motivation for `inplace` parameter?

Comment (lvhan028) [inline lmdeploy/pytorch/engine/logits_process.py:94]: Let's remove `inplace`, because no `False` value is passed

File: src/turbomind/models/llama/LlamaV2.cc [hunk 12]
```diff
@@ -360,13 +360,8 @@ void LlamaV2<T>::dynamicDecode(int*            token_ids,
...
+    const std::vector<std::string> optional_inputs{
+        "stop_words_list", "bad_words_list", "runtime_top_k", "runtime_top_p", "temperature", "repetition_penalty"};
```
[the remark this answers is not in the document]
Comment (irexyc) [reply src/turbomind/models/llama/LlamaV2.cc:364]: Most variables except `stop_words_list` and `bad_words_list` have already been initialized by `DynamicDecodeLayer::setup`.

File: src/turbomind/kernels/sampling_topp_kernels.cu [hunk 13]
```diff
@@ -27,795 +27,134 @@
...
-                                                               reinterpret_cast<Value_Data_Type*>(params.gmem_src_vals),
-                                                               reinterpret_cast<Value_Data_Type*>(params.gmem_dst_vals),
-                                                               params.num_items,
-                                                               params.num_segments,
-                                                               params.gmem_begin_offsets,
-                                                               params.gmem_end_offsets,
-                                                               0,
-                                                               sizeof(Key_Data_Type) * 8,
-                                                               stream);
+    for (int tid = threadIdx.x; tid < vocab_size_padded; tid += blockDim.x) {
```
Comment (lzhangzz) [inline src/turbomind/kernels/sampling_topp_kernels.cu:93]: Shall we skip this softmax for top-k request?

File: src/turbomind/layers/sampling_layers/SamplingLayer.cc [hunk 14]
```diff
@@ -0,0 +1,292 @@
...
+    //  - sampling from left and sorted scores
+
+    TM_LOG_DEBUG("%s start", __PRETTY_FUNCTION__);
+
+    Tensor    logits     = input_tensors->at("logits");
+    const int batch_size = logits.shape[0];
+    const int step       = input_tensors->at("step").getVal<int>();
+    logits_              = logits.getPtr<T>();
+
+    cudaAutoCpy(kept_, kept_n_.data(), batch_size, stream_);
```
[the remark this answers is not in the document]
Comment (irexyc) [reply src/turbomind/layers/sampling_layers/SamplingLayer.cc:161]: Do you mean `kept_`? `kept_` needs to be assigned `vocab_size` at each step

File: src/turbomind/kernels/sampling_topp_kernels.cu [hunk 15]
```diff
@@ -27,795 +27,134 @@
...
+    __syncthreads();
 
-    int index = tid + bid * blockDim.x;
-
-    while (index < batch_size * n) {
-        topp_id_val_buf[index] = index % n;
-        index += blockDim.x * gridDim.x;
+    sum_val = s_sum_val;
+    for (int tid = threadIdx.x; tid < vocab_size_padded; tid += blockDim.x) {
+        logits[tid] = ((float)logits[tid] / (sum_val + 1e-6f));
```
[the remark this answers is not in the document]
Comment (irexyc) [reply src/turbomind/kernels/sampling_topp_kernels.cu]: removed since sum_val will always greater than 0

File: src/turbomind/kernels/sampling_topp_kernels.cu (hunk 13 above)
Comment (irexyc) [reply src/turbomind/kernels/sampling_topp_kernels.cu]: removed

**Review by lvhan028 — APPROVED** [review]

**Review by lzhangzz — APPROVED** [review]

**Comment by josephrocca** [comment]
Very excited to test this - thank you! 🙏
